---
### Exploration of Ebay Car Sales Data
---

The aim of this project is to clean the data and analyze used car listings. 
Data is in the following columns:
* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [1]:
import pandas as pd
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

As shown above the data contains:
* 20 columns of features - 5 columns with null / missing values.
* 15 columns are mixed object type - 5 columns are int values - not the same 5 with null values.
* The dataset contains 20 columns, most of which are strings.
Some columns have null values, but none have more than ~20% null values.
* The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

In [3]:
auto_cols = autos.columns
auto_cols

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Changes:

* Convert the column names from camelcase to snakecase.
* Reword some column names based on the data dictionary to make them more descriptive.

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [6]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

columns with almost all the same value 
* seller, offer type, ab_test, num_photos

columns for further investigation are
* vehicle_type, gearbox, registration_year, fuel_type, odometer and price

In [7]:
autos.drop(['seller','offer_type','ab_test','num_photos'], axis = 1, inplace = True)

In [8]:
autos['vehicle_type'].value_counts()

limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64

In [9]:
autos['gearbox'].value_counts()

manuell      36993
automatik    10327
Name: gearbox, dtype: int64

In [10]:
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [11]:
autos['registration_year'].value_counts().sort_index()

1000       1
1001       1
1111       1
1500       1
1800       2
1910       9
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       5
1957       2
1958       4
1959       7
1960      34
1961       6
1962       4
1963       9
        ... 
2001    2703
2002    2533
2003    2727
2004    2737
2005    3015
2006    2708
2007    2304
2008    2231
2009    2098
2010    1597
2011    1634
2012    1323
2013     806
2014     666
2015     399
2016    1316
2017    1453
2018     492
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: registration_year, Length: 97, dtype: int64

In [12]:
autos['fuel_type'].value_counts()

benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [13]:
autos['price'].value_counts().sort_index()

$0             1421
$1              156
$1,000          639
$1,039            1
$1,040            1
$1,049            6
$1,050           95
$1,059            1
$1,070            1
$1,080            6
$1,090            4
$1,095            3
$1,098            1
$1,099           44
$1,100          376
$1,111           39
$1,112            1
$1,119            1
$1,120            2
$1,149           10
$1,150          226
$1,169            1
$1,170            1
$1,180            4
$1,189            1
$1,190           37
$1,195            1
$1,199          126
$1,200          639
$1,201            2
               ... 
$925              1
$93,000           2
$93,911           1
$930              4
$94,999           1
$940              2
$945              2
$949             11
$950            379
$951              1
$958              1
$960              1
$965              2
$970              7
$975              2
$98,500           1
$980             48
$985              4
$989              2


In [14]:
autos['ad_created'].value_counts().sort_index()

2015-06-11 00:00:00       1
2015-08-10 00:00:00       1
2015-09-09 00:00:00       1
2015-11-10 00:00:00       1
2015-12-05 00:00:00       1
2015-12-30 00:00:00       1
2016-01-03 00:00:00       1
2016-01-07 00:00:00       1
2016-01-10 00:00:00       2
2016-01-13 00:00:00       1
2016-01-14 00:00:00       1
2016-01-16 00:00:00       1
2016-01-22 00:00:00       1
2016-01-27 00:00:00       3
2016-01-29 00:00:00       1
2016-02-01 00:00:00       1
2016-02-02 00:00:00       2
2016-02-05 00:00:00       2
2016-02-07 00:00:00       1
2016-02-08 00:00:00       1
2016-02-09 00:00:00       2
2016-02-11 00:00:00       1
2016-02-12 00:00:00       3
2016-02-14 00:00:00       2
2016-02-16 00:00:00       1
2016-02-17 00:00:00       1
2016-02-18 00:00:00       2
2016-02-19 00:00:00       3
2016-02-20 00:00:00       2
2016-02-21 00:00:00       3
                       ... 
2016-03-09 00:00:00    1662
2016-03-10 00:00:00    1593
2016-03-11 00:00:00    1639
2016-03-12 00:00:00    1831
2016-03-13 00:00:00 

* The price column has 1421 zero values and 3 just under 1m, as well as the '$' character in each entry
* The registration year has a number of entries before cars were invented or in the future
* The majority of the odometer readings are over 150k km, the values also contain the 'km' string
* The ads were mostly posted between March and April 2016 but a small number were posted as far back as June 2015

In [15]:
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',' , '').astype(float)
autos['price'] = autos['price'].str.replace('$','').str.replace(',' , '').astype(float)
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price vehicle_type  registration_year    gearbox  power_ps   model  \
0  5000.0          bus               2004    manuell       158  andere   
1  8500.0    limousine               1997  automatik       286     7er   
2  8990.0    limousine               2009    manuell       102    golf   
3  4350.0   kleinwagen               2007  automatik        71  fortwo   
4  1350.0        kombi               2003    manuell         0   focus   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0     150000.0                   3       lpg     peugeot              nein   
1     150000.0                   6    benzin         bmw              nein   
2      70000.0                   7    benzin  volkswagen              nein   
3      70000.0                   6    benzin       smart              nein   
4     150000.0                   7    benzin        ford              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

Analyze the columns, using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

In [16]:
print('Price:\nShape: ', autos['price'].shape,'\n')
print('Stats:\n ', autos['price'].describe(),'\n')
print('Value counts:\n ', autos['price'].value_counts().sort_index(ascending = True),'\n')

Price:
Shape:  (50000,) 

Stats:
  count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

Value counts:
  0.0           1421
1.0            156
2.0              3
3.0              1
5.0              2
8.0              1
9.0              1
10.0             7
11.0             2
12.0             3
13.0             2
14.0             1
15.0             2
17.0             3
18.0             1
20.0             4
25.0             5
29.0             1
30.0             7
35.0             1
40.0             6
45.0             4
47.0             1
49.0             4
50.0            49
55.0             2
59.0             1
60.0             9
65.0             5
66.0             1
              ... 
151990.0         1
155000.0         1
163500.0         1
163991.0         1
169000.0         1
169999.0         1
175000.0         1
180000.0         1
19

The price column has a 14 entries over 100k and 1421 zero values, the 75% percentile is $7200 but it is reasonable to expect a a number of more expensive vehicles to be true data points.

In [17]:
print('Odometer:\nShape: ', autos['odometer_km'].shape,'\n')
print('Stats:\n ', autos['odometer_km'].describe(),'\n')
print('Value counts:\n ', autos['odometer_km'].value_counts().sort_index(ascending = True),'\n')

Odometer:
Shape:  (50000,) 

Stats:
  count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

Value counts:
  5000.0        967
10000.0       264
20000.0       784
30000.0       789
40000.0       819
50000.0      1027
60000.0      1164
70000.0      1230
80000.0      1436
90000.0      1757
100000.0     2169
125000.0     5170
150000.0    32424
Name: odometer_km, dtype: int64 



With the odometers the most common value is >150k, given the prices for the majority this tallies. 

In [18]:
autos = autos[autos['price'].between(1,350000)]
autos.describe()

price  registration_year      power_ps    odometer_km  \
count   48565.000000       48565.000000  48565.000000   48565.000000   
mean     5888.935591        2004.755421    117.197158  125770.101925   
std      9059.854754          88.643887    200.649618   39788.636804   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7490.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        48565.000000  48565.000000  
mean             5.782251  50975.745207  
std              3.685595  25746.968398  
min              0.000000   1067.000000  
25%              3.000000  30657.000000  
50%              6.000000  49716.000000  
75%              9.000000  71665.000000  
max             12.000000  99998.000000

* Registration year needs inspecting
* Power_ps cannot be 0

In [19]:
autos = autos[autos['power_ps'] > 0]
autos['registration_year'].value_counts().sort_index(ascending = True)

1800       2
1927       1
1929       1
1931       1
1937       3
1941       2
1950       2
1951       2
1954       2
1955       1
1956       3
1957       2
1958       2
1959       6
1960       9
1961       4
1962       3
1963       7
1964      10
1965      13
1966      13
1967      18
1968      24
1969      14
1970      19
1971      20
1972      29
1973      16
1974      21
1975      14
        ... 
1995    1014
1996    1203
1997    1743
1998    2105
1999    2569
2000    2550
2001    2415
2002    2277
2003    2491
2004    2497
2005    2590
2006    2537
2007    2148
2008    2103
2009    1996
2010    1543
2011    1588
2012    1275
2013     786
2014     643
2015     365
2016     876
2017    1046
2018     350
2019       2
2800       1
4800       1
5000       1
5911       1
9000       1
Name: registration_year, Length: 79, dtype: int64

There were no cars before ~1900, and as it's only 2021 the values greater must be dropped

In [20]:
autos = autos[autos['registration_year'].between(1900,2021)]
autos.describe()

price  registration_year      power_ps    odometer_km  \
count   43572.000000       43572.000000  43572.000000   43572.000000   
mean     6221.754269        2003.610530    130.527495  125591.893877   
std      9197.129555           7.177885    207.345012   39342.641069   
min         1.000000        1927.000000      1.000000    5000.000000   
25%      1380.000000        1999.000000     80.000000  100000.000000   
50%      3449.500000        2004.000000    116.000000  150000.000000   
75%      7900.000000        2008.000000    150.000000  150000.000000   
max    350000.000000        2019.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        43572.000000  43572.000000  
mean             5.957450  51475.841114  
std              3.589622  25740.228851  
min              0.000000   1067.000000  
25%              3.000000  31139.000000  
50%              6.000000  50668.000000  
75%              9.000000  72352.000000  
max             12.000000  99998.000000

In [21]:
autos.tail()

date_crawled                                               name  \
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

         price vehicle_type  registration_year    gearbox  power_ps   model  \
49995  24900.0    limousine               2011  automatik       239      q5   
49996   1980.0       cabrio               1996    manuell        75   astra   
49997  13200.0       cabrio               2014  automatik        69     500   
49998  22900.0        kombi               2013    manuell       150      a3   
49999   1250.0    limousine               1996    manuell       101  vectra   

       odometer_km  registration_month fuel_type brand unrepaired_damage  \
49995     100000.0                   1    diesel  audi              nein   
49996     150000.0                   5    benzin  opel              nein   
49997       5000.0                  11    benzin  fiat              nein   
49998      40000.0                  11    diesel  audi              nein   
49999     150000.0                   1    benzin  opel              nein   

                ad_created  postal_code            last_seen  
49995  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999  2016-03-13 00:00:00        45897  2016-04-06 21:18:48

#### Date columns

In [22]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2016-03-05    0.025406
2016-03-06    0.014183
2016-03-07    0.036285
2016-03-08    0.033209
2016-03-09    0.032659
2016-03-10    0.032154
2016-03-11    0.032062
2016-03-12    0.037249
2016-03-13    0.015629
2016-03-14    0.036698
2016-03-15    0.034196
2016-03-16    0.029468
2016-03-17    0.031396
2016-03-18    0.012944
2016-03-19    0.034426
2016-03-20    0.038350
2016-03-21    0.037249
2016-03-22    0.032796
2016-03-23    0.031626
2016-03-24    0.029354
2016-03-25    0.031786
2016-03-26    0.032429
2016-03-27    0.030960
2016-03-28    0.035459
2016-03-29    0.033554
2016-03-30    0.033645
2016-03-31    0.031947
2016-04-01    0.033875
2016-04-02    0.035527
2016-04-03    0.039039
2016-04-04    0.036767
2016-04-05    0.013128
2016-04-06    0.003190
2016-04-07    0.001354
Name: date_crawled, dtype: float64

These dates are fairly evenly spread between March and April 2016

In [23]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)

2015-08-10    0.000023
2015-09-09    0.000023
2015-11-10    0.000023
2015-12-05    0.000023
2015-12-30    0.000023
2016-01-03    0.000023
2016-01-07    0.000023
2016-01-10    0.000046
2016-01-13    0.000023
2016-01-14    0.000023
2016-01-16    0.000023
2016-01-22    0.000023
2016-01-27    0.000069
2016-01-29    0.000023
2016-02-01    0.000023
2016-02-02    0.000046
2016-02-05    0.000046
2016-02-07    0.000023
2016-02-08    0.000023
2016-02-09    0.000023
2016-02-12    0.000046
2016-02-14    0.000023
2016-02-16    0.000023
2016-02-17    0.000023
2016-02-18    0.000046
2016-02-19    0.000046
2016-02-20    0.000046
2016-02-21    0.000069
2016-02-22    0.000023
2016-02-23    0.000069
                ...   
2016-03-09    0.032727
2016-03-10    0.031924
2016-03-11    0.032360
2016-03-12    0.037088
2016-03-13    0.017006
2016-03-14    0.035344
2016-03-15    0.033990
2016-03-16    0.029905
2016-03-17    0.031167
2016-03-18    0.013564
2016-03-19    0.033301
2016-03-20    0.038396
2016-03-21 

Again, these seem relatiely even and realistic in spread

In [24]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index(ascending = True)


2016-03-05    0.001033
2016-03-06    0.004108
2016-03-07    0.005003
2016-03-08    0.006954
2016-03-09    0.009341
2016-03-10    0.010121
2016-03-11    0.012049
2016-03-12    0.023432
2016-03-13    0.008583
2016-03-14    0.012508
2016-03-15    0.015652
2016-03-16    0.016019
2016-03-17    0.027449
2016-03-18    0.007184
2016-03-19    0.015767
2016-03-20    0.020655
2016-03-21    0.020036
2016-03-22    0.020931
2016-03-23    0.018521
2016-03-24    0.019393
2016-03-25    0.019026
2016-03-26    0.016410
2016-03-27    0.015170
2016-03-28    0.020426
2016-03-29    0.021941
2016-03-30    0.024832
2016-03-31    0.023341
2016-04-01    0.022928
2016-04-02    0.025085
2016-04-03    0.024832
2016-04-04    0.024167
2016-04-05    0.127123
2016-04-06    0.226063
2016-04-07    0.133916
Name: last_seen, dtype: float64

These dates also line up quite well with the others from initial review

In [25]:
autos['date_crawled'] = pd.to_datetime(autos['date_crawled'].str[:10])
autos['ad_created'] = pd.to_datetime(autos['ad_created'].str[:10])
autos['last_seen'] = pd.to_datetime(autos['last_seen'].str[:10])

In [26]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'chrysler', 'renault',
       'audi', 'sonstige_autos', 'opel', 'mazda', 'porsche', 'mini',
       'ford', 'mercedes_benz', 'seat', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'land_rover', 'alfa_romeo', 'rover',
       'daihatsu', 'trabant', 'lancia', 'lada'], dtype=object)

In [27]:
autos['brand'].value_counts().sort_values(ascending = False)

volkswagen        9322
bmw               4919
opel              4652
mercedes_benz     4165
audi              3878
ford              2979
renault           1963
peugeot           1275
fiat              1087
seat               839
skoda              733
mazda              672
nissan             640
smart              607
citroen            602
toyota             555
hyundai            443
volvo              406
mini               404
mitsubishi         356
honda              355
sonstige_autos     342
kia                311
alfa_romeo         293
porsche            265
suzuki             256
chevrolet          243
chrysler           153
dacia              119
jeep                98
daihatsu            97
subaru              94
land_rover          91
saab                74
jaguar              66
daewoo              57
rover               52
lancia              47
trabant             41
lada                21
Name: brand, dtype: int64

I will include all brands in the aggregation of mean price 

In [28]:
mean_prices = {}
brands = autos['brand'].unique()

for i in brands:
    x = autos[autos['brand'] == i]
    y = x['price'].mean()
    mean_prices[i] = int(y)
mean_prices

{'alfa_romeo': 4130,
 'audi': 9608,
 'bmw': 8532,
 'chevrolet': 6596,
 'chrysler': 3583,
 'citroen': 3893,
 'dacia': 6071,
 'daewoo': 1079,
 'daihatsu': 1760,
 'fiat': 2964,
 'ford': 3904,
 'honda': 4184,
 'hyundai': 5624,
 'jaguar': 12063,
 'jeep': 12036,
 'kia': 6132,
 'lada': 3021,
 'lancia': 3406,
 'land_rover': 19665,
 'mazda': 4291,
 'mercedes_benz': 8920,
 'mini': 10600,
 'mitsubishi': 3529,
 'nissan': 5073,
 'opel': 3119,
 'peugeot': 3227,
 'porsche': 48507,
 'renault': 2619,
 'rover': 1767,
 'saab': 3341,
 'seat': 4598,
 'skoda': 6641,
 'smart': 3721,
 'sonstige_autos': 13828,
 'subaru': 4041,
 'suzuki': 4340,
 'toyota': 5291,
 'trabant': 2013,
 'volkswagen': 5621,
 'volvo': 5069}

Porsche has the highest mean at 48500
Landrover has the next highest at 19600
Sonstige are surprising at 13800, higher than Jeep & Jaguar at 12000

In [29]:
mean_km = {}
brands = autos['brand'].unique()

for i in brands:
    x = autos[autos['brand'] == i]
    y = x['odometer_km'].mean()
    mean_km[i] = int(y)
mean_km

{'alfa_romeo': 131126,
 'audi': 128933,
 'bmw': 132674,
 'chevrolet': 97489,
 'chrysler': 133790,
 'citroen': 119343,
 'dacia': 83025,
 'daewoo': 124122,
 'daihatsu': 117113,
 'fiat': 117143,
 'ford': 124348,
 'honda': 122169,
 'hyundai': 105112,
 'jaguar': 127424,
 'jeep': 127806,
 'kia': 112443,
 'lada': 85952,
 'lancia': 125531,
 'land_rover': 118076,
 'mazda': 124799,
 'mercedes_benz': 130710,
 'mini': 89207,
 'mitsubishi': 125702,
 'nissan': 117148,
 'opel': 129210,
 'peugeot': 126678,
 'porsche': 97716,
 'renault': 127200,
 'rover': 136730,
 'saab': 143243,
 'seat': 121185,
 'skoda': 110156,
 'smart': 98904,
 'sonstige_autos': 94912,
 'subaru': 125478,
 'suzuki': 108730,
 'toyota': 115711,
 'trabant': 55487,
 'volkswagen': 128466,
 'volvo': 138706}

In [30]:
cost = pd.Series(mean_prices)
dist = pd.Series(mean_km)
df = pd.DataFrame(cost, columns = ['mean_price'])
df['mean_km'] = dist
df['mean_cost_per_km'] = round(1000 * df['mean_price'] / df['mean_km'],2)
df = df.sort_values(by = ['mean_cost_per_km'], ascending = False)

df

mean_price  mean_km  mean_cost_per_km
porsche              48507    97716            496.41
land_rover           19665   118076            166.55
sonstige_autos       13828    94912            145.69
mini                 10600    89207            118.82
jaguar               12063   127424             94.67
jeep                 12036   127806             94.17
audi                  9608   128933             74.52
dacia                 6071    83025             73.12
mercedes_benz         8920   130710             68.24
chevrolet             6596    97489             67.66
bmw                   8532   132674             64.31
skoda                 6641   110156             60.29
kia                   6132   112443             54.53
hyundai               5624   105112             53.50
toyota                5291   115711             45.73
volkswagen            5621   128466             43.75
nissan                5073   117148             43.30
suzuki                4340   108730             39.92
seat                  4598   121185             37.94
smart                 3721    98904             37.62
volvo                 5069   138706             36.54
trabant               2013    55487             36.28
lada                  3021    85952             35.15
mazda                 4291   124799             34.38
honda                 4184   122169             34.25
citroen               3893   119343             32.62
subaru                4041   125478             32.20
alfa_romeo            4130   131126             31.50
ford                  3904   124348             31.40
mitsubishi            3529   125702             28.07
lancia                3406   125531             27.13
chrysler              3583   133790             26.78
peugeot               3227   126678             25.47
fiat                  2964   117143             25.30
opel                  3119   129210             24.14
saab                  3341   143243             23.32
renault               2619   127200             20.59
daihatsu              1760   117113             15.03
rover                 1767   136730             12.92
daewoo                1079   124122              8.69

The most expensive vehicles in the data set are;

* The porsche at an average 496 per 1000km on the clock.
* Followed by Land Rover, Songstie and Mini all between 100 - 200 per 1000km.
* all other brands are between the 15 - 100 per 1000km.

The best value per mile done are the Daewoo, Rover and Daihatsu models, all coming in at 15 per 1000km.

The most standout brands are Songstie, Dacia and Cheverolet which all come in at a higher cost per 1000km than BMW, given BWM tend to be at more of a premium than these brands.

---
#### Final remarks
---

n this project we practiced applying a variety of pandas methods to explore and understand a data set on car listings. 

Data cleaning next steps:
* Identify categorical data that uses german words, translate them and map the values to their english counterparts
* Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
* See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
* Find the most common brand/model combinations
* Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
* How much cheaper are cars with damage than their non-damaged counterparts?